### ISSUE TITLE
Stray percentage sign (%) in dynamic segments of URLs cause 500

### ISSUE THREAD

**Issue by klarstrup** [vercel/next.js#43303]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information
    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.8.0
      npm: 8.18.0
      Yarn: N/A
      pnpm: 7.14.1
    Relevant packages:
      next: 13.0.5-canary.6
      eslint-config-next: 13.0.4
      react: 18.2.0
      react-dom: 18.2.0

### Describe the Bug

A percentage sign not followed by two hex digits in a dynamic path segment will trigger a 500 Internal Server Error on the frontend, and in development mode an error with a stacktrace of this sort:

```
Error: failed to decode param
    at decode (/Users/io/code/vizsla/node_modules/next/dist/shared/lib/router/utils/route-matcher.js:17:23)
    at /Users/io/code/vizsla/node_modules/next/dist/shared/lib/router/utils/route-matcher.js:27:21
    at Array.forEach (<anonymous>)
    at Object.match (/Users/io/code/vizsla/node_modules/next/dist/shared/lib/router/utils/route-matcher.js:21:29)
    at DevServer.runMiddleware (/Users/io/code/vizsla/node_modules/next/dist/server/next-server.js:1093:50)
    at async DevServer.runMiddleware (/Users/io/code/vizsla/node_modules/next/dist/server/dev/next-dev-server.js:644:28)
    at async Object.fn (/Users/io/code/vizsla/node_modules/next/dist/server/next-server.js:1201:38)
    at async Router.execute (/Users/io/code/vizsla/node_modules/next/dist/server/router.js:252:36)
    at async DevServer.run (/Users/io/code/vizsla/node_modules/next/dist/server/base-server.js:383:29)
    at async DevServer.run (/Users/io/code/vizsla/node_modules/next/dist/server/dev/next-dev-server.js:734:20)
```

I've found two public production websites exhibiting this bug, from the Next.js showcase website list:

https://institchu.com/design/% (This one forwards the stacktrace!)

https://futurism.com/%

We've also run into it over at my company: https://etilbudsavis.dk/% (which brought me here, as this `500` reports with a high severity in our logs)

### Expected Behavior

I would expect a `400 Bad Request` error rather than a `500 Internal Server Error` in this scenario, as this is a malform URL.

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://github.com/vercel/next.js/tree/canary/examples/dynamic-routing

### To Reproduce

Visit one of these sites running various Next.js versions:

~https://etilbudsavis.dk/% (Next.js version 13.2.4 at time of writing)~

https://institchu.com/design/% (Next.js version 10.2.3 at time of writing)

https://futurism.com/% (Next.js version 12.1.0 at time of writing)

**Comment by koba04** [comment]
I couldn't reproduce it, and the test case seems to be passed.
https://github.com/vercel/next.js/blob/canary/test/integration/dynamic-routing/test/index.test.js#L1117-L1120

**Comment by klarstrup** [comment]
You are right, this seems to have been fixed somewhere between 12.1.0 and 13.2.4 by my testing just now. Curiously that test is much older than that.
